### ISSUE TITLE
Script Doesn't Properly Refresh the Currently Hooked Window

### ISSUE THREAD

**Issue by xuru-thor** [padiix/RecORDER#26]
Pretty much every game I play won't have their recordings stored in their own specific folders, they will only be moved to the "Manual Recording" directory. This happens because I always play in borderless windowed mode, instead of fullscreen.
This could be fixed by having the script move the recordings to folders named after literally any window that was currently open when the recording was saved, regardless if it was in fullscreen mode or not.
This could create some clutter when it comes to desktop recordings, but I believe this to be a way better solution than borderless windowed games not being properly recognized. Maybe it could be an additional optional setting?

**Comment by padiix** [comment]
Hello!

I am not sure, but I don't think the Borderless Fullscreen is the issue here, because if the OBS can capture the screen without issues, so would my script.
Everything in it uses nothing except data from the OBS itself.

Have you followed the set-up and F.A.Q on main page of this repository?
The cause for your problem might be there.

Other than that, please provide logs from the "Script Logs" window in the Script window after experiencing your issue.
If you do not I cannot help you!

**Comment by xuru-thor** [comment]
Sorry, I don't think I was clear enough. The recording and the plugin are working the way they should. The problem is recordings and screenshots are only saved in the generic "Manual Recording" folder from the plugin, instead of specific folders for each application. I assume this to be an issue with borderless windowed mode, since you detailed in the readme description that this is a feature only present to fullscreen applications (and hooked windows, but i only use display capture in OBS).

**Comment by padiix** [comment]
### High Level explaination:
Well, there is the problem here - You are using the **Display Capture** in OBS, so it doesn't really know **what you are capturing at this moment**. 

Because of that, there is no need for the OBS to know what type of window/fullscreen application he is capturing because he is capturing the whole screen.
In the consequence, my script has no idea what folder to put your videos/screenshots, because OBS itself doesn't know.


### Lower Level explaination (programist pov to say it easier)
My script makes use of the **Signals** (think of a specific events, that need to be set-up in specific time of script execution) that the OBS provides for scripts to make sense of what is being **"hooked"** (in this particular case; hooked as in what OBS can capture while using the Game Capture or Window Capture).

Here's the signal I'm talking about:
![image](https://github.com/user-attachments/assets/16066dda-dcc7-4645-b60d-3630f8346bba)

As you can see, **only a few elements are making use of it**, because in case of the others, the OBS does not necessarily need to hook itself to the game screen/window.

Because of that, I think it is hardly a matter of the script or the OBS, as it is a matter of the source (as in the one you selected in OBS) that **you are using**.

### TLDR
**I am fairly certain that this is not OBS or Script issue.**
You are using the **Display Capture source** which causes OBS to not provide necessary information for my script to work.
Try using the **Window Capture** or **Game Capture** at your own choice (_I think Window Capture will be better if you prefer the Display Capture_).

If I had to make the Display Capture work as well, then my script would need additional libraries, which would make usage of my script more painful and against what I wanted to make it.

**Let me know if my recommendation worked and if not - try recording a brief moment and share logs in here!**

**Comment by xuru-thor** [comment]
Yes, Window Capture orders the recordings perfectly (with Red Dead Redemption 2, at least. The Witcher still seems to get saved to the Manual Recording folder, I've attached a log for it here).

![image_2024-12-29_092305539](https://github.com/user-attachments/assets/e47be5be-99e6-429e-b6c8-e937af199f0d)

However, I've found the Window Capture source to be extremely lacking compared to Display Capture, as it wouldn't always correctly capture the window that was actually open, or some of the task bar windows. BUT, adding the Display Capture source on top of the Window Capture Source actually fixes this issue completely, while retaining properly organized recordings!
The only issue I have right now is Witcher not being hooked by OBS, causing recordings from it to be saved in the generic directory. Perhaps the Window Capture option is not correctly refreshing its window hooks? I'm not sure, am hoping you'd know.
I'll also test this setup with more games to see if there are any other issues. Thanks a lot for the attention and help.

**Comment by padiix** [comment]
Sometimes the problem with hooks is fixed by re-applying the script (deletion and addition basically), but yea, I agree that the Window Capture can be lacking, because it is fixed on whatever the screen is and sometimes the title changes during game, which breaks it.

So generally the best source for gameplay is the Game Capture, but Window Capture still works for games that cannot be hooked normally by the OBS.

**Comment by xuru-thor** [comment]
I've attempted to use Game Capture just now, and I am 100% it's hooked to The Witcher window. Yet, the script still moves it to the Manual Recording folder... This happens with Minecraft as well.
I tried reapplying the script as well, but that didn't seem to have changed anything.

**Comment by padiix** [comment]
Hmm, interesting...

What version of OBS are you using?
Do you change the sources names in any way?
What language do you have your system in?
Do you have any other scripts running in the background?

**Comment by xuru-thor** [comment]
Using the latest version, 30.2.3.
No, I use the default preset names, "Display Capture" and "Game Capture".
My system is running in English (United States).
No, this is the only script I'm running.

**Comment by xuru-thor** [comment]
Noticed something interesting. When selecting the "Capture specific window" option with Game Capture, The Witcher is not present in the dropdown menu options at all. It's almost like OBS does not detect it as it's own window.

**Comment by padiix** [comment]
That's interesting indeed, but we wouldn't know if this is an OBS issue, since you are able to hook onto the window.

Latest version of OBS available is 31.0.0, not 30.2.3, but I guess you might be on Linux.
![image](https://github.com/user-attachments/assets/e06b2cf6-f899-4bd4-83ce-7b550442887e)



What Python version are you using in the OBS?
You can check it here:
![image](https://github.com/user-attachments/assets/071ea389-5634-4519-8dee-43708a8bad05)

**Comment by xuru-thor** [comment]
I'm using Python version 3.11, the one you recommended. However, I am on Windows, and it turns out I really was using an outdated version of OBS. I never updated it since I never got a pop up to do so. Updating it to 31.0.0 immediately fixed the issue with The Witcher window not hooking, and also fixed the issue of recordings not being saved to its own specific folder LMFAO

Although, now I am experiencing a different issue... It seems like the script isn't properly refreshing the currently hooked window. Booting up The Witcher, taking a screenshot, closing the game, then booting up Red Dead Redemption, and taking a screenshot there, will cause said screenshot to be saved inside The Witcher recording directory. I have attached a screenshot of the log here.

![image_2024-12-30_174726994](https://github.com/user-attachments/assets/9c917bf2-9df8-4f8a-b43a-521aba7c82e3)

**Comment by padiix** [comment]
Good to hear about the first part!

About the second part - Did you re-apply the script after updating?
Sometimes the update breaks the scripts (no idea why)

**Comment by xuru-thor** [comment]
I had not reapplied the script, did that just now. Issue persists, unfortunately.

**Comment by padiix** [comment]
Peculiar... 
Could you show how you have the Python setup in the Scripts window?

I am also assuming, that you use the 1.5.1 version of my script, right?

**Comment by xuru-thor** [comment]
![image_2024-12-30_213151952](https://github.com/user-attachments/assets/edc71080-86fb-4c3b-a425-976ef439b8ec)

Yes, I'm using 1.5.1, the latest release.

**Comment by padiix** [comment]
This is very suprising, I'll try to look at it, maybe I forgot to put a source UUID clear somewhere in my code.

I'll try to work on version 1.6 soon, because I was stalling it due to it only having some optimization and other not really needed stuff.

**Comment by xuru-thor** [comment]
Thank you for all the work you're putting into this project, I wish you good luck on fixing this issue.

**Comment by padiix** [comment]
Alright, so my thoughts were good as I think they would - The script works for me, I downloaded the 1.5.1 version that is shared on this repository to make sure that this will be a fair check.
Just as you I also used two games, here are my logs:

```
[RecORDER.py] (script_update) Updated the settings!
[RecORDER.py] 
[RecORDER.py] (script_update) Updated the settings!
[RecORDER.py] 
[RecORDER.py] (script_update) Updated the settings!
[RecORDER.py] 
[RecORDER.py] []--------------------------[]
[RecORDER.py] Reloading the signals...
[RecORDER.py] <>--------------------------<>
[RecORDER.py] Checking available sources for a match with source table...
[RecORDER.py] Match found!
[RecORDER.py] <>--------------------------<>
[RecORDER.py] 
[RecORDER.py] Signals reloaded.
[RecORDER.py] Running get_hooked procedure to get current app title...
[RecORDER.py] Checking if source is hooked to any window...
[RecORDER.py] Hooked!
[RecORDER.py] Current game title: Unpacking
[RecORDER.py] User took the screenshot...
[RecORDER.py] []--------------------------[]
[RecORDER.py] 
[Unknown Script] >remember_and_move<>remember_and_move<>remember_and_move<
[Unknown Script] (Asyncio) Done!
[Unknown Script] (Asyncio) File moved to: B:\Kreatywne\YouTube\Captures\Unpacking\Screenshots\Screenshot 241231_10-47-34.png
[Unknown Script] >remember_and_move<>remember_and_move<>remember_and_move<
[Unknown Script] 
[RecORDER.py] <<>>--------------------------<<>>
[RecORDER.py] Fetching data from calldata...
[RecORDER.py] gameTitle: SIGNALIS
[RecORDER.py] <<>>--------------------------<<>>
[RecORDER.py] <<>>--------------------------<<>>
[RecORDER.py] Fetching data from calldata...
[RecORDER.py] gameTitle: SIGNALIS
[RecORDER.py] <<>>--------------------------<<>>
[RecORDER.py] []--------------------------[]
[RecORDER.py] User took the screenshot...
[RecORDER.py] []--------------------------[]
[RecORDER.py] 
[Unknown Script] >remember_and_move<>remember_and_move<>remember_and_move<
[Unknown Script] (Asyncio) Done!
[Unknown Script] (Asyncio) File moved to: B:\Kreatywne\YouTube\Captures\SIGNALIS\Screenshots\Screenshot 241231_10-49-31.png
[Unknown Script] >remember_and_move<>remember_and_move<>remember_and_move<
[Unknown Script]
```

As you can see, the script correctly picked up on the source changes and so it also took care of the organizing.

My queries to you:
* Do you have multiple sources in your current scene/scene collection?
* Could you share a screenshot of your normal main OBS screen, so I can make sure that other elements are also correct?


I will still check the code, but from what I experienced now it looks like you have some very edge case scenario on your end that I did not anticipate.

**Comment by padiix** [comment]
Alright, checked the code and I understood what's happening. I messily did some code for screenshot handling without really thinking about it much more.


![image](https://github.com/user-attachments/assets/f4681d24-7b30-4944-bd4b-4d2cf6b89927)

From the gist of the code above:
1. Code checks if it already has the UUID of source in memory, if not, then it will fetch it.
2. Code checks if the Title of captured screen is different than the default one - if yes, then it fetches again to have the current one.

The problem is at step 2.
Because it ONLY checks the title on the situation where the Title is the same as "Manual Recording", it creates a situation where the Title is the same if the signal for hooking doesn't fire, like in your case.

To put it simply, because the "hooked" signal did not work as it should, the screenshot was placed in the same place as the previous one.

Signals are sometimes unreliable, but only way to get the window title correctly.

**Comment by xuru-thor** [comment]
Wow, extremely interesting that you've found an error in your code, as I was just now typing up how I've found a solution.

Yes, I was running more than one source in OBS. A Display Capture source, so that I wouldn't get a blank display on my recordings whenever viewing a non-fullscreen window or my desktop, and a Game Capture source beneath it, so that windows would properly hook with RecORDER. This was the issue, and hooked windows would properly update upon deleting the Display Capture source.
And, it turns out, Display Capture can be kept without breaking the hooking function. All you have to do is place that source BELOW Game Capture, and everything will work as it's supposed to.

**Comment by padiix** [comment]
It's not really the code error, it's more of a logic error - I sometimes forgot that the signals are not to be relied upon, because they are moody at some points, so I just removed the part of the code that might create it.

As for the source order - it's very important to keep it clean to avoid issues like that, so good of you to figure it out yourself!

I will try to release the version 1.6 today, just need to make sure everything makes sense in it.

**Comment by padiix** [comment]
Alright then, since everything seems to be working for you and I found the error that could possibly cause issues in the future, I will close the issue.

Feel free to open it if you still find another issues related to your problems today!
